### ISSUE TITLE
ARC optimization: 10 experiments to beat DSPy 91.3%

### ISSUE THREAD

**Issue by jkim-fsai** [jkim-fsai/AFlow#24]
## Goal

Beat DSPy's best ARC score of **91.3%** (GEPA #8, gpt-4.1 coach, heavy budget).

## Context

**DSPy reference:** 89.5% baseline → 91.3% best (+1.8%), $9.75 total, ~69 min across 8 experiments.

**AFlow experiment 1 results:** 88.0% baseline → 90.6% best (+2.6%), $1.95, ~30 min.
- Best rounds used simple Custom operator (CoT or CoT+Review)
- 5/11 rounds used ScEnsemble which **always** regressed to 62-66%
- ScEnsemble's XML output biases toward "D" regardless of question

## Experiments

All experiments use `--dataset ARC --mode Graph --exec_model_name gpt-4.1-nano`.

Baseline accuracy: **88.0%** (CoT prompt, gpt-4.1-nano, 299-example validation set)

| # | Coach Model | Max Rounds | Sample | Val Rounds | Workspace | Rationale |
|---|-------------|------------|--------|------------|-----------|-----------|
| 1 | gpt-4.1 | 10 | 6 | 2 | workspace_arc_v1 | **Done.** Best 90.6% |
| 2 | gpt-4.1-nano | 10 | 6 | 2 | workspace_arc_e2 | Cheap coach baseline |
| 3 | gpt-4.1 | 20 | 6 | 2 | workspace_arc_e3 | 2x exploration budget |
| 4 | gpt-4.1 | 10 | 3 | 2 | workspace_arc_e4 | Smaller sample → more MCTS exploration |
| 5 | gpt-4.1 | 10 | 10 | 2 | workspace_arc_e5 | Larger sample → more MCTS exploitation |
| 6 | gpt-4.1-nano | 20 | 3 | 2 | workspace_arc_e6 | Cheap coach + many rounds + exploration |
| 7 | gpt-4.1 | 15 | 6 | 1 | workspace_arc_e7 | More rounds, faster validation |
| 8 | gpt-4.1-nano | 15 | 10 | 2 | workspace_arc_e8 | Cheap coach + exploitation-heavy |
| 9 | gpt-4.1 | 30 | 6 | 2 | workspace_arc_e9 | Heavy budget (3x rounds) |
| 10 | gpt-4.1 | 20 | 4 | 2 | workspace_arc_e10 | Mid-range balanced |

## Setup per experiment

Each workspace must be bootstrapped before running:
```bash
mkdir -p workspace_arc_eN/ARC/workflows
cp -r workspace/ARC/workflows/template workspace_arc_eN/ARC/workflows/
cp -r workspace/ARC/workflows/round_1 workspace_arc_eN/ARC/workflows/
```

### Command template
```bash
uv run python run.py \
  --dataset ARC --mode Graph \
  --opt_model_name <coach> \
  --exec_model_name gpt-4.1-nano \
  --max_rounds <rounds> \
  --sample <sample> \
  --validation_rounds <val_rounds> \
  --optimized_path <workspace>
```

## Results table (to be filled)

| # | Coach | Rounds | Sample | Val | Optimized | Delta | Cost | Runtime |
|---|-------|--------|--------|-----|-----------|-------|------|---------|
| 1 | gpt-4.1 | 10 | 6 | 2 | 90.6% | +2.6% | ~$1.95 | ~30 min |
| 2 | | | | | | | | |
| 3 | | | | | | | | |
| 4 | | | | | | | | |
| 5 | | | | | | | | |
| 6 | | | | | | | | |
| 7 | | | | | | | | |
| 8 | | | | | | | | |
| 9 | | | | | | | | |
| 10 | | | | | | | | |
| | | | | | | **Total** | | |

## Acceptance criteria

- [ ] All 10 experiments completed with results recorded
- [ ] Results table updated with all scores, costs, and runtimes
- [ ] Best AFlow score identified and compared to DSPy's 91.3%
- [ ] Observations documented (which hyperparameters matter most)

## Known issues

- **ScEnsemble regression**: The optimizer frequently generates ScEnsemble (3x voting) workflows that always regress to ~65%. This wastes ~50% of optimization rounds. A fix to restrict operators for ARC could significantly improve results.
- **Workspace bootstrapping**: New workspaces must have `template/` and `round_1/` copied in before running.
- **File deletion bug**: Do NOT use `--optimized_path workspace` — it deletes all round files. Use unique workspace names.

---
*GJ65MV6QP1 @ /Users/joshuakim/Github/aa/AFlow*

**Comment by jkim-fsai** [comment]
## Code change for Experiment 3+

**Removed ScEnsemble and AnswerGenerate from ARC operator set** in `run.py`:
```python
# Before (experiments 1-2):
operators=["Custom", "AnswerGenerate", "ScEnsemble"]

# After (experiment 3+):
operators=["Custom"]
```

**Rationale:** ScEnsemble regressed to 62-66% in every round it was used (5/11 in exp 1, 8/11 in exp 2). AnswerGenerate produces prose answers incompatible with ARC's single-letter format. Restricting to Custom-only ensures every optimization round is productive.

This is a fair comparison — DSPy's GEPA/MIPROv2 optimizers only optimize prompts, not workflow topology. AFlow's Custom operator is the equivalent.

*GJ65MV6QP1 @ /Users/joshuakim/Github/aa/AFlow*

**Comment by jkim-fsai** [comment]
## Experiment 4 Complete

**Config:** gpt-4.1 coach, 10 rounds, sample 6, val 2, Custom-only
**Best: 90.6%** | Cost: $1.34 | Runtime: ~16 min

**Key finding — controlled experiment answers the question:**
- Exp 1 (All operators, 10 rounds): 90.6%
- Exp 4 (Custom-only, 10 rounds): 90.6% ← same ceiling
- Exp 3 (Custom-only, 20 rounds): 93.6% ← breakthrough

**Both the operator fix AND extra rounds were needed.** Custom-only removes wasted rounds but 10 rounds isn't enough exploration to discover the 93%+ prompt patterns. Exp 3's rounds 3-14 found multi-step CoT refinements that 10 rounds missed.

Note: Round 6 hit 36% catastrophically even with Custom-only — the optimizer can still generate bad prompts, just less frequently.

### Updated results table

| # | Coach | Rounds | Sample | Val | Operators | Optimized | Delta | Cost | Runtime |
|---|-------|--------|--------|-----|-----------|-----------|-------|------|---------|
| 1 | gpt-4.1 | 10 | 6 | 2 | All | 90.6% | +2.6% | ~$1.95 | ~30 min |
| 2 | gpt-4.1-nano | 10 | 6 | 2 | All | 90.0% | +2.0% | ~$3.78 | ~45 min |
| 3 | gpt-4.1 | 20 | 6 | 2 | Custom | **93.6%** | **+5.6%** | ~$2.27 | ~22 min |
| 4 | gpt-4.1 | 10 | 6 | 2 | Custom | 90.6% | +2.6% | ~$1.34 | ~16 min |

*GJ65MV6QP1 @ /Users/joshuakim/Github/aa/AFlow*

**Comment by jkim-fsai** [comment]
## Experiment 5 Complete — Nano matches gpt-4.1!

**Config:** gpt-4.1-nano coach, 20 rounds, sample 6, val 2, Custom-only
**Best: 93.6%** | Cost: $1.44 | Runtime: ~19 min | Converged at round 13

**Key finding:** With Custom-only operators, nano coach matches gpt-4.1 (both 93.6%). And cheaper ($1.44 vs $2.27). The gpt-4.1 advantage in exps 1-2 was only that it generated fewer ScEnsemble rounds — remove that, and nano is equally capable.

Round 11 hit 22.9% catastrophically but the optimizer recovered. Both coaches occasionally produce bad prompts.

### Updated results table

| # | Coach | Rounds | Sample | Val | Operators | Optimized | Delta | Cost | Runtime |
|---|-------|--------|--------|-----|-----------|-----------|-------|------|---------|
| 1 | gpt-4.1 | 10 | 6 | 2 | All | 90.6% | +2.6% | ~$1.95 | ~30 min |
| 2 | gpt-4.1-nano | 10 | 6 | 2 | All | 90.0% | +2.0% | ~$3.78 | ~45 min |
| 3 | gpt-4.1 | 20 | 6 | 2 | Custom | **93.6%** | **+5.6%** | ~$2.27 | ~22 min |
| 4 | gpt-4.1 | 10 | 6 | 2 | Custom | 90.6% | +2.6% | ~$1.34 | ~16 min |
| 5 | gpt-4.1-nano | 20 | 6 | 2 | Custom | **93.6%** | **+5.6%** | ~$1.44 | ~19 min |

*GJ65MV6QP1 @ /Users/joshuakim/Github/aa/AFlow*

**Comment by jkim-fsai** [comment]
## Experiment 7 Complete — Converged early, 91.3%

**Config:** gpt-4.1 coach, 30 rounds, sample 3, val 2, Custom-only
**Best: 91.3%** | Cost: $1.67 | Runtime: ~30 min | Converged at round 15/30

**Surprising finding:** gpt-4.1 coach + sample=3 underperformed nano coach + sample=3 (91.3% vs 94.3%). The gpt-4.1 coach converged too early to a local optimum. The nano coach's more random exploration with sample=3 produced greater diversity and found better prompts in later rounds.

### Updated results table

| # | Run ID | Coach | Rounds | Sample | Val | Ops | Optimized | Delta | Cost | Runtime |
|---|--------|-------|--------|--------|-----|-----|-----------|-------|------|---------|
| 1 | workspace_arc_v1 | gpt-4.1 | 10 | 6 | 2 | All | 90.6% | +2.6% | ~$1.95 | ~30 min |
| 2 | workspace_arc_e2 | gpt-4.1-nano | 10 | 6 | 2 | All | 90.0% | +2.0% | ~$3.78 | ~45 min |
| 3 | workspace_arc_e3 | gpt-4.1 | 20 | 6 | 2 | Custom | 93.6% | +5.6% | ~$2.27 | ~22 min |
| 4 | workspace_arc_e4 | gpt-4.1 | 10 | 6 | 2 | Custom | 90.6% | +2.6% | ~$1.34 | ~16 min |
| 5 | workspace_arc_e5 | gpt-4.1-nano | 20 | 6 | 2 | Custom | 93.6% | +5.6% | ~$1.44 | ~19 min |
| 6 | workspace_arc_e6 | gpt-4.1-nano | 20 | 3 | 2 | Custom | **94.3%** | **+6.3%** | ~$4.82 | ~46 min |
| 7 | workspace_arc_e7 | gpt-4.1 | 30 | 3 | 2 | Custom | 91.3% | +3.3% | ~$1.67 | ~30 min |

*GJ65MV6QP1 @ /Users/joshuakim/Github/aa/AFlow*
